coordinator/src/main/scala/filodb.coordinator/TenantIngestionMetering.scala

```diff
@@ -16,6 +17,92 @@ import filodb.coordinator.client.QueryCommands.LogicalPlan2Query
 import filodb.core.DatasetRef
 import filodb.query.{QueryError, QueryResult, TsCardinalities}
 
+object QueryThrottle {
+  // currently just add this diff to the interval if the timeout rate exceeds THRESHOLD
+  protected val INTERVAL_DIFF = FiniteDuration(5L, TimeUnit.MINUTES)
+  // number of past query timeouts/non-timeouts to consider
+  protected val LOOKBACK = 10
+  // {non-timeouts-in-lookback-window} / LOOKBACK < THRESHOLD will adjust the interval
+  protected val THRESHOLD = 0.85
+}
+
+object TenantIngestionMetering {
+  protected val METRIC_ACTIVE = "active_timeseries_by_tenant"
+  protected val METRIC_TOTAL = "total_timeseries_by_tenant"
+}
+
+/**
+ * Throttles the TenantIngestionMetering query rate according to the ratio of timeouts to non-timeouts.
+ *
+ * @param queryInterval the initial delay between each query. This is the duration to be adjusted.
+ */
+class QueryThrottle(queryInterval: FiniteDuration) extends StrictLogging {
+  import QueryThrottle._
+
+  private var interval: FiniteDuration = queryInterval.copy()
+  private val lock = new ReentrantReadWriteLock()
```

Comment:
Exists only because there may be more than one dataset (see `queryAndSchedule`).

Reply:
removed